api/rtx/src/uvisor_malloc.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <stdint.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <reent.h>
+
+#include "uvisor-lib/uvisor-lib.h"
+#include "cmsis_os.h"
+
+#define DPRINTF(...) {};
+/* Use printf with caution inside malloc: printf may allocate memory itself,
+   so using printf in malloc may lead to recursive calls! */
+/* #define DPRINTF(fmt, ...) printf(fmt, ## __VA_ARGS__) */
+
+/* Declare this variable here, so uVisor initializes enough space for the
+ * mutexes, which are used in this file! */
+const uint32_t __uvisor_box_index_size = sizeof(RtxBoxIndex);
+
+extern RtxBoxIndex *const __uvisor_ps;
+extern void uvisor_malloc_init(void);
+
+static int is_kernel_initialized()
+{
+    static uint8_t kernel_running = 0;
+    return (kernel_running || (kernel_running = osKernelRunning()));
+}
+
+static int init_allocator()
+{
+    int ret = 0;
+    if (__uvisor_ps == NULL) {
+#if defined(UVISOR_PRESENT) && (UVISOR_PRESENT == 1)
+        return -1;
+#else
+        uvisor_malloc_init();
+#endif
+    }
+
+    if ((__uvisor_ps->mutex_id == NULL) && is_kernel_initialized()) {
+        /* point the mutex pointer to the data */
+        __uvisor_ps->mutex.mutex = &(__uvisor_ps->mutex_data);
+        /* create mutex if not already done */
+        __uvisor_ps->mutex_id = osMutexCreate(&(__uvisor_ps->mutex));
+        /* mutex failed to be created */
+        if (__uvisor_ps->mutex_id == NULL) {
+            return -1;
+        }
+    }
+
+    if (__uvisor_ps->index.active_heap == NULL) {
+        /* we need to initialize the process heap */
+        if (__uvisor_ps->index.box_heap != NULL) {
+            /* lock the mutex during initialization */
+            int kernel_initialized = is_kernel_initialized();
```

Comment:
If `is_kernel_initialized` changes types, change the type of `kernel_initialized` here as well.